core/sr-primitives/src/lib.rs

```diff
@@ -688,22 +753,28 @@ macro_rules! impl_outer_log {
 		}
 
 		impl From<$crate::generic::DigestItem<$($genarg),*>> for $name {
-			/// Converts `generic::DigestItem` into `$name`. If `generic::DigestItem` represents
-			/// a system item which is supported by the runtime, it is returned.
-			/// Otherwise we expect a `Other` log item. Trying to convert from anything other
-			/// will lead to panic in runtime, since the runtime does not supports this 'system'
-			/// log item.
+			/// Converts `generic::DigestItem` into `$name`. If
+			/// `generic::DigestItem` represents a system item which is
+			/// supported by the runtime, it is returned. Otherwise we expect a
+			/// `Other`, `PreDigest`, or `Consensus` log item. Trying to convert
+			/// from anything else will lead to panic at runtime, since the
+			/// runtime does not supports this 'system' log item.
 			#[allow(unreachable_patterns)]
 			fn from(gen: $crate::generic::DigestItem<$($genarg),*>) -> Self {
 				match gen {
 					$($(
-					$crate::generic::DigestItem::$sitem(value) =>
-						$name($internal::$module($module::RawLog::$sitem(value))),
+					$crate::__parse_pattern!($sitem b, a) =>
+						$name($crate::__parse_expr!($sitem b, $module $internal a)),
 					)*)*
-					_ => gen.as_other()
+					_ => {
+						if let Some(s) = gen.as_other()
 						.and_then(|value| $crate::codec::Decode::decode(&mut &value[..]))
-						.map($name)
-						.expect("not allowed to fail in runtime"),
+						.map($name) {
+							s
+						} else {
+							panic!("Unhandled digest in runtime")
+						}
+					}
```

Comment:
What is protecting this from panicking?

Reply:
The reason for avoiding `expect` is so that the line number displayed in case of a panic points to the correct place, rather than somewhere in `libcore`.

Reply:
> What is protecting this from panicking?

^^^

Reply:
@gavofyork @marcio-diaz This can only panic if the runtime is bad and does not handle a digest that it receives.  This means the runtime is broken, which is fatal.

Reply:
Isn't it possible to receive a bad digest?

Reply:
@marcio-diaz The old code also `panic`ed here (though with a worse error message), so this is not a regression.  If the bad digest comes from a block created by another node, the block will (correctly) be rejected.  Otherwise, the bad digest had to come from the runtime itself or from the consensus engine.  Both are fatal errors.

Reply:
> The old code also paniced here (though with a worse error message), so this is not a regression.

I know ^^.

>  If the bad digest comes from a block created by another node, the block will (correctly) be rejected

For curiosity, where is this logic? `check_header`? Because I'm not so sure in that case.